Correctly update segment index max_offset after truncation

ztlpn opened the pull request:
## Cover letter

Previously after truncation max_offset was only updated if we had to modify the index. So if truncation happened near the end of the segment, it wasn't updated, leading to segments overlap after redpanda restarted. This overlap then manifests in readers unexpectedly stopping before they reach the end of their range.

We also detect situations when after restart we get a segments overlap (this could happen if e.g. we failed to write the segment index file) and fix them by rebuilding the problematic indexes.

Fixes: #2478

jcsp commented on src/v/raft/offset_translator.cc:
```diff
@@ -394,6 +394,14 @@ ss::future<> offset_translator::sync_with_log(
       log_consumer{*this}, model::no_timeout);
 
+    if (_highest_known_offset < log_offsets.dirty_offset) {
+        throw std::runtime_error{_logger.format(
```
Perhaps this should be a vassert?  I think this assertion would go up to the controller_backend when it's starting a partition & cause that to retry continuously, which would be hard for a user to diagnose/understand.

ztlpn replied:
Well, maybe! I'm still not sure about what is the right policy for choosing between assertion/exception. It seems unfair to crash the whole server due to problems with one partition, but OTOH the user indeed has almost no way to determine what happened.

jcsp commented:
I wonder if this also fixes https://github.com/vectorizedio/redpanda/issues/2144, which had a similar set of circumstances (loading indices with bad max_offsets)

jcsp commented:
This is really nice work.

For reference, I dug up a couple of never-merged commits that I was using when investigating #2144 here: https://github.com/vectorizedio/redpanda/pull/3133

I think your changes to segment_set are a superset of the case that my logging was covering in that patch.

Maybe worth pulling this assertion (https://github.com/vectorizedio/redpanda/pull/3133/commits/cc05d3ab15d18afffc8fb9a297cd5a9f5adc6f83) into this PR as an extra sanity check?  Makes the tests fail more cleanly if we have a too-far-ahead index.

jcsp commented on src/v/raft/offset_translator.cc:
```diff
@@ -394,6 +394,14 @@ ss::future<> offset_translator::sync_with_log(
     co_await std::move(reader).for_each_ref(
       log_consumer{*this}, model::no_timeout);
 
+    if (_highest_known_offset < log_offsets.dirty_offset) {
```
This assumes that the log had at least one batch in it, right? (otherwise _highest_known_offset would be in its initial ::min() state).  That's usually true but would be nice not to rely on other components only trying to sync with non-empty segments.

ztlpn replied:
Actually, offset_translator goes through a rather convoluted startup sequence where it is first `start`-ed, then `prefix_truncate_reset` is performed with data from the raft snapshot, and then it is `sync_with_log`-ed. So before the sync `_highest_known_offset` is at least `_last_snapshot_index`. But I'll check with empty logs.

ztlpn replied:
Yeah, for a new partition highest known offset is -infinity and dirty offset is -infinity as well, and if we have no segments, but a valid snapshot, highest known offset is equal to `_last_snapshot_index` which is also the dirty offset.

jcsp commented on src/v/storage/segment_set.cc:
```diff
@@ -159,27 +160,55 @@ unsafe_do_recover(segment_set&& segments, ss::abort_source& as) {
+            if (i > 0) {
+                auto& prev = *good[i - 1];
+
+                if (prev.offsets().dirty_offset >= s.offsets().base_offset) {
```
I think we can make this a more comprehensive check by also checking that dirty_offset >= max_offset here, in case of other offset bug -- it's a great position to check all the things we possibly can, while are still early enough in startup to rewrite indices.

ztlpn replied:
> also checking that dirty_offset >= max_offset here

Can you elaborate? I thought they are essentially the same, i.e. dirty_offset gets assigned index.max_offset after the index is hydrated.

jcsp replied:
Right, they should be the same - I'm basically suggesting some paranoia: this is a convenient place to check all the invariants we assume about the state of the index/segment variables.

ztlpn commented:
changes in the [force-push](https://github.com/vectorizedio/redpanda/compare/3d04dbb5fea0edc39ce09812df1aab58d9ad1427..a11c74c9390dc36322e5b057964d1e69bac06c6c):
* addressed review feedback
* fixed tests - turns out archival tests create partitions rather crudely - by manually creating some segments (with gaps between them) and then calling `partition_mgr.manage()`. The segments didn't start from offset 0 which is inconsistent in several ways - e.g. not consistent with the log `start_offset` saved in kvstore, there is no snapshot etc. The right way to fix it is probably to insert the data in a proper way and then do a prefix-truncate but simply starting the first segment from offset 0 seems to do the trick.
